src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs

```diff
@@ -0,0 +1,337 @@
+                {
+                    Name = "expires_at",
+                    Value = expiresAt.ToString("o", CultureInfo.InvariantCulture)
+                });
+            }
+
+            if (!string.IsNullOrEmpty(tokens.TokenType))
+            {
+                tokensToStore.Add(new AuthenticationToken
+                {
+                    Name = "token_type",
+                    Value = tokens.TokenType
+                });
+            }
+
+            properties.StoreTokens(tokensToStore);
+        }
+
+        var identity = new ClaimsIdentity(ClaimsIssuer);
+        var ticket = await CreateTicketAsync(identity, properties, tokens);
+        return HandleRequestResult.Success(ticket);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+    {
+        // Both device_id and code_verifier are required to get access token
+        if (!context.Properties.Items.TryGetValue(VkIdAuthenticationConstants.AuthenticationProperties.DeviceId, out var deviceId) ||
+            string.IsNullOrEmpty(deviceId))
+        {
+            return OAuthTokenResponse.Failed(new Exception("Device ID was not found."));
+        }
+
+        if (!context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier) ||
+            string.IsNullOrEmpty(codeVerifier))
+        {
+            return OAuthTokenResponse.Failed(new Exception("Code verifier key was not found."));
+        }
+
+        context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        var query = new Dictionary<string, string>
+        {
+            ["grant_type"] = "authorization_code",
+            ["code"] = context.Code,
+            ["code_verifier"] = codeVerifier,
+            ["client_id"] = Options.ClientId,
+            ["device_id"] = deviceId,
+            ["redirect_uri"] = context.RedirectUri,
+            ["state"] = Options.StateDataFormat.Protect(context.Properties),
+        };
+
+        using var request = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        request.Content = new FormUrlEncodedContent(query);
+
+        var response = await Backchannel.SendAsync(request, Context.RequestAborted);
+
+        // According to docs even error response should be 200
+        if (response.StatusCode is not HttpStatusCode.OK)
+        {
+            await Log.ExchangeCodeErrorAsync(Logger, response, Context.RequestAborted);
+            return OAuthTokenResponse.Failed(new Exception("Invalid remote server response during code exchange."));
+        }
+
+        var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
+
+        // Code exchange response should always contain state parameter.
+        if (!payload.RootElement.TryGetProperty("state", out var state) ||
```

Comment:
`state` in token responses is not really a standard thing and it's not clear what security benefits validating the `state` here provides. I'd just remove that part to simplify things.